File: app/src/main/java/tpcreative/co/qrscanner/common/ScannerSingleton.java
package tpcreative.co.qrscanner.common;

public class ScannerSingleton {
    private static ScannerSingleton instance ;
    private SingletonScannerListener listener;

    public static ScannerSingleton getInstance(){
        if (instance==null){
            synchronized (ScannerSingleton.class){
                if (instance==null){
                    instance = new ScannerSingleton();
                }
            }
        }
        return instance;
    }

    public void setListener(SingletonScannerListener listener){
        this.listener = listener;
    }

    public void setVisible(){
        if (listener!=null){
            listener.setVisible();
        }
    }

    public void setInvisible(){
        if (listener!=null){
            listener.setInvisible();
        }
    }

    public interface SingletonScannerListener{
        void setVisible();
        void setInvisible();
    }

}

File: app/src/main/java/tpcreative/co/qrscanner/common/ScannerSingleton.kt
package tpcreative.co.qrscanner.common

class ScannerSingleton {
    private var listener: SingletonScannerListener? = null
    fun setListener(listener: SingletonScannerListener?) {
        this.listener = listener
    }

    fun setVisible() {
        if (listener != null) {
            listener?.setVisible()
        }
    }

    fun setInvisible() {
        if (listener != null) {
            listener?.setInvisible()
        }
    }

    interface SingletonScannerListener {
        fun setVisible()
        fun setInvisible()
    }

    companion object {
        private var instance: ScannerSingleton? = null
        fun getInstance(): ScannerSingleton? {
            if (instance == null) {
                synchronized(ScannerSingleton::class.java) {
                    if (instance == null) {
                        instance = ScannerSingleton()
                    }
                }
            }
            return instance
        }
    }
}